### PR TITLE
TF2 code for chapter 17 (GANs)

### PR DESCRIPTION
**Pull request by biswajitsahoo1111**
This pull request is to merge TF2 code for section 17.1 of GAN chapter. After its merger functions like `d2l.update_D` and `d2l.update_G` will be available in the module code of TF that can be used in section 17.2.

PS: While going through past pull requests, I found [1716](https://github.com/d2l-ai/d2l-en/pull/1716) which was very similar. This pull request is not a copy of that one. Most notable difference, among other things, is the scaled version of loss function that matches with PyTorch's loss value. 

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

**Comment by mli** [comment]
Job d2l-en/PR-1760/1 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1760/

**Comment by mli** [comment]
Job d2l-en/PR-1760/6 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1760/

File: chapter_generative-adversarial-networks/dcgan.md [hunk 1]
```diff
@@ -126,6 +163,19 @@ class G_block(nn.Module):
         return self.activation(self.batch_norm(self.conv2d_trans(X)))
 ```
 
+```{.python .input}
+#@tab tensorflow
+class G_block(tf.keras.layers.Layer):
+    def __init__(self, out_channels, kernel_size = 4, strides = 2, padding = "same", **kwargs):
```
Comment (terrytangyuan) [inline chapter_generative-adversarial-networks/dcgan.md:169]: Remove redundant spaces around equal signs

File: chapter_generative-adversarial-networks/dcgan.md [hunk 2]
```diff
@@ -167,6 +224,13 @@ g_blk = G_block(20, strides=1, padding=0)
 g_blk(x).shape
 ```
 
+```{.python .input}
+#@tab tensorflow
+x = tf.zeros((2, 1, 1, 3))
+g_blk = G_block(20, strides = 1, padding= "valid") # `padding = "valid"` corresponds to no padding
```
Comment (terrytangyuan) [inline chapter_generative-adversarial-networks/dcgan.md:230]: Same here and other places

**Review by terrytangyuan — APPROVED** [review]
Just some nit picks. Otherwise LGTM. Thanks!

**Review by terrytangyuan — APPROVED** [review]

**Comment by mli** [comment]
Job d2l-en/PR-1760/7 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1760/

File: chapter_generative-adversarial-networks/dcgan.md [hunk 3]
```diff
@@ -407,16 +543,71 @@ def train(net_D, net_G, data_iter, num_epochs, lr, latent_dim,
...
 
+```{.python .input}
+#@tab pytorch
+latent_dim, lr, num_epochs = 100, 0.005, 20
+train(net_D, net_G, data_iter, num_epochs, lr, latent_dim)
+```
+
+```{.python .input}
+#@tab tensorflow
+latent_dim, lr, num_epochs = 100, 0.0005, 40  # We have changed the learning rate and number of epochs
```
Comment (astonzhang) [inline chapter_generative-adversarial-networks/dcgan.md:607]: Why do we choose a different set of HPs?

**Comment by astonzhang** [comment]
@biswajitsahoo1111 Thanks!

Just minor comments:
1. We can use "#@tab mxnet, pytorch" to avoid repeating the same code block.
2. We keep each line <= 78 chars for a consistent style 
3. For consistency, we prefer single quotes (padding= "valid" -> padding= 'valid')

Don't worry. I'll fix these!

File: chapter_generative-adversarial-networks/dcgan.md (hunk 3 above)
Comment (biswajitsahoo1111) [reply chapter_generative-adversarial-networks/dcgan.md:607]: I tried several times using the shared learning rate and number of epochs. But I couldn't get any reasonable looking generated images in any of the trials. However, if we change the learning rate and number of epochs to the values that we have used (0.0005, and 40 respectively), almost always we get reasonable looking generated images. That's the reason why I changed those for TF.

**Comment by biswajitsahoo1111** [comment]
> @biswajitsahoo1111 Thanks!
> 
> Just minor comments (I've just partially fixed them):
> 
> 1. We can use "#@tab mxnet, pytorch" to avoid repeating the same code block.
> 2. We keep each line <= 78 chars for a consistent style
> 3. For consistency, we prefer single quotes (padding= "valid" -> padding= 'valid')
> 
> More on the style can be found at https://github.com/d2l-ai/d2l-en/blob/master/STYLE_GUIDE.md

@astonzhang Thanks. I was not aware of point 1. Otherwise, I would have used that. I would be more careful about point 2 and 3.

File: chapter_generative-adversarial-networks/dcgan.md (hunk 3 above)
Comment (astonzhang) [reply chapter_generative-adversarial-networks/dcgan.md:607]: Can you get similar results to those produced by other frameworks when using the same HPs?
